File: scripts/mini-bench-run.py
#!/usr/bin/env python3
"""
mini-bench-run.py - Run 20 queries against pre-indexed corpus and report score.

FTS queries: test keyword/phrase retrieval edge cases
Semantic queries: test embedding-based recall

Expected result: substring that should appear in at least one top-5 slug or title.
"""

import argparse
import json
import os
import subprocess
import sys
import time

# ─── Query definitions ────────────────────────────────────────────────────────
# Each entry: (query_string, expected_substring_in_slug_or_title, label)

FTS_QUERIES = [
    ("agent memory architecture",        "Agent Memory",        "S01"),
    ("stablecoin regulation",            "Stablecoin Reg",      "S02"),
    ("DeFi liquidity protocol",          "DeFi",                "S03"),
    ("Rust async runtime performance",   "Rust",                "S04"),
    ("vector embedding search",          "vector",              "S05"),
    ("knowledge graph traversal",        "Knowledge Graph",     "S06"),
    ("smart contract security audit",    "Smart Contract",      "S07"),
    ("cross-chain bridge mechanism",     "Cross-chain",         "S08"),
    ("zero knowledge proof circuit",     "Zero Knowledge",      "S09"),
    ("retrieval augmented generation",   "Retrieval",           "S10"),
]

SEMANTIC_QUERIES = [
    ("how do AI agents remember information across multiple sessions",
     "Agent Memory",   "Q01"),
    ("what makes decentralized exchanges different from centralized ones",
     "DeFi",           "Q02"),
    ("techniques for compressing large language model context windows",
     "LLM",            "Q03"),
    ("regulatory challenges for stablecoin issuers in the US",
     "Stablecoin",     "Q04"),
    ("building efficient search over large markdown document collections",
     "vector",         "Q05"),
    ("how to detect contradictions in a knowledge base",
     "Knowledge",      "Q06"),
    ("performance optimisation strategies for Rust web services",
     "Rust",           "Q07"),
    ("blockchain interoperability and cross-chain asset transfers",
     "Cross-chain",    "Q08"),
    ("cryptographic commitments and privacy in smart contracts",
     "Zero Knowledge", "Q09"),
    ("graph-based reasoning and link traversal in document stores",
     "Knowledge Graph", "Q10"),
]


# ─── Runner ───────────────────────────────────────────────────────────────────

def run_query(quaid_bin: str, db_path: str, query: str,
              cmd: str = "search", limit: int = 5) -> list[dict]:
    env = {**os.environ, "QUAID_DB": db_path}
    result = subprocess.run(
        [quaid_bin, cmd, query, "--limit", str(limit), "--json"],
        env=env, capture_output=True, text=True, timeout=15,
    )
    if result.returncode != 0:
        return []
    try:
        data = json.loads(result.stdout)
        return data if isinstance(data, list) else data.get("results", [])
    except Exception:
        return []


def check_hit(results: list[dict], expected: str) -> bool:
    for r in results:
        slug = r.get("slug", "").lower()
        title = r.get("title", "").lower()
        if expected.lower() in slug or expected.lower() in title:
            return True
    return False


def bar(n: int, total: int, width: int = 10) -> str:
    filled = round(n / total * width)
    return "█" * filled + "░" * (width - filled)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--db", default="/tmp/quaid-mini-bench.db")
    parser.add_argument("--quaid", default="./target/release/quaid")
    parser.add_argument("--prev", default="/tmp/quaid-mini-bench-prev.txt")
    args = parser.parse_args()

    quaid = os.path.abspath(args.quaid)
    if not os.path.isfile(quaid):
        print(f"ERROR: quaid binary not found at {quaid}")
        print("Run: cargo build --release")
        sys.exit(1)

    if not os.path.isfile(args.db):
        print(f"ERROR: bench DB not found at {args.db}")
        print("Run: python3 scripts/mini-bench-setup.py")
        sys.exit(1)

    # Load previous score
    prev_score = None
    if os.path.isfile(args.prev):
        try:
            prev_score = int(open(args.prev).read().strip())
        except Exception:
            pass

    fts_results = []
    sem_results = []
    t_start = time.time()

    # FTS
    for query, expected, label in FTS_QUERIES:
        hits = run_query(quaid, args.db, query, cmd="search")
        passed = check_hit(hits, expected)
        fts_results.append((label, passed))

    # Semantic
    for query, expected, label in SEMANTIC_QUERIES:
        hits = run_query(quaid, args.db, query, cmd="query")
        passed = check_hit(hits, expected)
        sem_results.append((label, passed))

    elapsed = time.time() - t_start

    fts_pass = sum(1 for _, p in fts_results if p)
    sem_pass = sum(1 for _, p in sem_results if p)
    total = fts_pass + sem_pass
    max_total = 20

    fts_labels = " ".join(f"{l}{'✓' if p else '✗'}" for l, p in fts_results)
    sem_labels = " ".join(f"{l}{'✓' if p else '✗'}" for l, p in sem_results)

    print()
    print(f"§3 FTS  [{bar(fts_pass, 10)}] {fts_pass}/10   {fts_labels}")
    print(f"§4 Sem  [{bar(sem_pass, 10)}] {sem_pass}/10   {sem_labels}")

    delta_str = ""
    if prev_score is not None:
        delta = total - prev_score
        sign = "+" if delta >= 0 else ""
        delta_str = f"  prev: {prev_score}/{max_total}  delta: {sign}{delta}"

    grade = "✅" if total >= 14 else ("⚠️ " if total >= 10 else "🔴")
    print(f"Total: {total}/{max_total}{delta_str}  ({elapsed:.1f}s)  {grade}")
    print()

    # Save score
    try:
        with open(args.prev, "w") as f:
            f.write(str(total))
    except Exception:
        pass

    # Regression gate: exit 1 if below threshold
    if total < 8:
        print("🔴 REGRESSION: score below gate (8/20)")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/mini-bench-run.mjs
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const REPO_DIR = path.dirname(__dirname);
const DEFAULT_DB = "/tmp/quaid-mini-bench.db";
const DEFAULT_QUAID = path.join(REPO_DIR, "target", "release", "quaid");
const DEFAULT_PREV = "/tmp/quaid-mini-bench-prev.txt";

const FTS_QUERIES = [
  { query: "agent memory architecture persistent storage sessions", expected: ["Agent Memory", "agent-memory", "page-0089"], label: "S01" },
  { query: "vector embeddings semantic search dense representations", expected: ["Vector Embeddings", "vector-embeddings", "vector"], label: "S02" },
  { query: "stablecoin regulation clarity institutional adoption", expected: ["Stablecoin Regulation", "stablecoin-regulation", "stablecoin"], label: "S03" },
  { query: "Rust async runtime performance optimization", expected: ["Rust Performance", "rust-performance", "rust"], label: "S04" },
  { query: "automated market maker DeFi liquidity protocol pooled reserves", expected: ["DeFi Liquidity", "defi-liquidity", "defi"], label: "S05" },
  { query: "knowledge graph traversal entity relationships", expected: ["Knowledge Graph", "knowledge-graph", "knowledge"], label: "S06" },
  { query: "smart contract security audit formal verification exploit", expected: ["Smart Contract", "smart-contract", "smart"], label: "S07" },
  { query: "cross-chain bridge interoperability asset transfer", expected: ["Cross-Chain", "cross-chain", "cross"], label: "S08" },
  { query: "zero knowledge proof cryptographic circuit commitment", expected: ["Zero Knowledge", "zero-knowledge", "zero"], label: "S09" },
  { query: "retrieval augmented generation RAG document search", expected: ["Retrieval Augmented", "rag-retrieval", "retrieval"], label: "S10" },
];

const SEMANTIC_QUERIES = [
  { query: "how do AI agents remember information across multiple sessions", expected: ["Agent Memory", "agent-memory", "page-0084", "page-0089", "page-0020"], label: "Q01" },
  { query: "what makes decentralized exchanges different from centralized ones", expected: ["DeFi", "defi-liquidity", "defi", "page-0046"], label: "Q02" },
  { query: "techniques for compressing large language model context windows", expected: ["Retrieval", "rag-retrieval", "retrieval", "page-0072"], label: "Q03" },
  { query: "regulatory challenges for stablecoin issuers in the US", expected: ["Stablecoin", "stablecoin-regulation", "stablecoin", "page-0009"], label: "Q04" },
  { query: "building efficient search over large markdown document collections", expected: ["Vector Embeddings", "vector-embeddings", "vector", "page-0038"], label: "Q05" },
  { query: "how to detect contradictions in a knowledge base graph", expected: ["Knowledge Graph", "knowledge-graph", "knowledge", "page-0034"], label: "Q06" },
  { query: "performance optimisation strategies for Rust web services", expected: ["Rust Performance", "rust-performance", "rust", "page-0044"], label: "Q07" },
  { query: "blockchain interoperability and cross-chain asset transfers", expected: ["Cross-Chain", "cross-chain", "cross", "page-0000"], label: "Q08" },
  { query: "cryptographic commitments and privacy in smart contracts", expected: ["Zero Knowledge", "zero-knowledge", "zero", "page-0055"], label: "Q09" },
  { query: "graph-based reasoning and link traversal in document stores", expected: ["Knowledge Graph", "knowledge-graph", "knowledge", "page-0071"], label: "Q10" },
];

function parseArgs(argv) {
  const args = {
    db: DEFAULT_DB,
    quaid: DEFAULT_QUAID,
    prev: DEFAULT_PREV,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--db") {
      args.db = argv[++i];
    } else if (arg === "--quaid") {
      args.quaid = argv[++i];
    } else if (arg === "--prev") {
      args.prev = argv[++i];
    } else {
      throw new Error(`unknown argument: ${arg}`);
    }
  }

  return args;
}

function runQuery(quaid, db, command, query) {
  const result = spawnSync(quaid, ["--db", db, command, query, "--limit", "5", "--json"], {
    cwd: REPO_DIR,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    timeout: 15_000,
  });

  if (result.status !== 0) {
    return [];
  }

  try {
    const data = JSON.parse(result.stdout);
    return Array.isArray(data) ? data : data.results || [];
  } catch {
    return [];
  }
}

function checkHit(results, expected) {
  const needles = Array.isArray(expected) ? expected : [expected];
  return results.some((result) => {
    const haystack = `${result.slug || ""} ${result.title || ""}`.toLowerCase();
    return needles.some((needle) => haystack.includes(needle.toLowerCase()));
  });
}

function bar(n, total, width = 10) {
  const filled = Math.round((n / total) * width);
  return `${"█".repeat(filled)}${"░".repeat(width - filled)}`;
}

function formatLabels(results) {
  return results.map(({ label, passed }) => `${label}${passed ? "✓" : "✗"}`).join(" ");
}

function readPrev(prevPath) {
  try {
    const raw = fs.readFileSync(prevPath, "utf8").trim();
    const parsed = Number.parseInt(raw, 10);
    return Number.isNaN(parsed) ? null : parsed;
  } catch {
    return null;
  }
}

const args = parseArgs(process.argv.slice(2));
const quaid = path.resolve(args.quaid);

if (!fs.existsSync(quaid)) {
  console.error(`ERROR: quaid binary not found at ${quaid}`);
  console.error("Run: cargo build --release");
  process.exit(1);
}

if (!fs.existsSync(args.db)) {
  console.error(`ERROR: bench DB not found at ${args.db}`);
  console.error("Run: node scripts/mini-bench-setup.mjs");
  process.exit(1);
}

const prevScore = readPrev(args.prev);
const start = performance.now();

const ftsResults = FTS_QUERIES.map((item) => {
  const results = runQuery(quaid, args.db, "search", item.query);
  return { label: item.label, passed: checkHit(results, item.expected) };
});

const semanticResults = SEMANTIC_QUERIES.map((item) => {
  const results = runQuery(quaid, args.db, "query", item.query);
  return { label: item.label, passed: checkHit(results, item.expected) };
});

const elapsedSeconds = (performance.now() - start) / 1000;
const ftsPass = ftsResults.filter((result) => result.passed).length;
const semanticPass = semanticResults.filter((result) => result.passed).length;
const total = ftsPass + semanticPass;
const maxTotal = 20;
const grade = total >= 14 ? "✅" : total >= 10 ? "⚠️" : "🔴";

let delta = "";
if (prevScore !== null) {
  const diff = total - prevScore;
  delta = `  prev: ${prevScore}/${maxTotal}  delta: ${diff >= 0 ? "+" : ""}${diff}`;
}

console.log("");
console.log(`§3 FTS  [${bar(ftsPass, 10)}] ${ftsPass}/10   ${formatLabels(ftsResults)}`);
console.log(`§4 Sem  [${bar(semanticPass, 10)}] ${semanticPass}/10   ${formatLabels(semanticResults)}`);
console.log(`Total: ${total}/${maxTotal}${delta}  (${elapsedSeconds.toFixed(1)}s)  ${grade}`);
console.log("");

try {
  fs.writeFileSync(args.prev, String(total), "utf8");
} catch {
  // Best-effort delta tracking should not fail the benchmark.
}

if (total < 8) {
  console.error("🔴 REGRESSION: score below gate (8/20)");
  process.exit(1);
}
